### PR TITLE
Document behaviour of `SubscribableListener` on rejection

### PR DESCRIPTION
**Pull request by DaveCTurner**
It's possible that the `executor` passed in to a `SubscribableListener`
rejects the task it should execute. This commit adds to the Javadocs a
description of what happens if so.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java [hunk 1]
```diff
@@ -191,6 +191,10 @@ public final void addListener(ActionListener<T> listener) {
      *                      <li>Pass the desired executor in as {@code executor}, and</li>
      *                      <li>Invoke {@link #addListener} using that executor.</li>
      *                      </ul>
+     *                      <p>
+     *                      If {@code executor} rejects the execution of the completion of the subscribing listener then the result is
+     *                      discarded and the subscribing listener is completed with a rejection exception on the thread which completes
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java:196]:
```suggestion
     *                      discarded and the subscribing listener is completed with an execution exception on the thread which completes
```

File: server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java [hunk 2]
```diff
@@ -451,6 +455,11 @@ public <U> SubscribableListener<U> andThen(CheckedBiConsumer<ActionListener<U>,
      * <li>Pass the desired executor in as {@code executor}, and</li>
      * <li>Invoke {@link #andThen} using that executor.</li>
      * </ul>
+     * <p>
+     * If {@code executor} rejects the execution of {@code nextStep} then the result is discarded and the returned listener is completed
+     * with a rejection exception on the thread which completes this listener. Likewise if this listener is completed exceptionally but
+     * {@code executor} rejects the execution of the completion of the returned listener then the returned listener is completed with a
+     * rejection exception on the thread which completes this listener.
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java:462]:
```suggestion
     * execution exception on the thread which completes this listener.
```

File: server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java [hunk 3]
```diff
@@ -451,6 +455,11 @@ public <U> SubscribableListener<U> andThen(CheckedBiConsumer<ActionListener<U>,
      * <li>Pass the desired executor in as {@code executor}, and</li>
      * <li>Invoke {@link #andThen} using that executor.</li>
      * </ul>
+     * <p>
+     * If {@code executor} rejects the execution of {@code nextStep} then the result is discarded and the returned listener is completed
+     * with a rejection exception on the thread which completes this listener. Likewise if this listener is completed exceptionally but
+     * {@code executor} rejects the execution of the completion of the returned listener then the returned listener is completed with a
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java:461]:
```suggestion
     * {@code executor} rejects the execution of the completion of the returned listener then the returned listener is completed with an
```

File: server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java [hunk 4]
```diff
@@ -191,6 +191,10 @@ public final void addListener(ActionListener<T> listener) {
      *                      <li>Pass the desired executor in as {@code executor}, and</li>
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java:191]: This line is unclear. I see elsewhere comments say `Ensure that this {@link SubscribableListener} is always completed using that executor, and` -- that would be helpful to update here and L455, IIUC.

File: server/src/test/java/org/elasticsearch/action/support/SubscribableListenerTests.java [hunk 5]
```diff
@@ -607,6 +610,32 @@ public void testAndThenAcceptFailure() {
         assertComplete(chainedListener, "simulated");
     }
 
+    public void testRejection() {
```
Comment (DiannaHohensee) [inline server/src/test/java/org/elasticsearch/action/support/SubscribableListenerTests.java:613]:
```suggestion
    public void testExecutorExecutionErrorPropagation() {
```

File: server/src/test/java/org/elasticsearch/action/support/SubscribableListenerTests.java [hunk 6]
```diff
@@ -607,6 +610,32 @@ public void testAndThenAcceptFailure() {
...
+        final Executor executor = r -> asInstanceOf(AbstractRunnable.class, r).onRejection(
+            new EsRejectedExecutionException("simulated rejection", randomBoolean())
+        );
+
+        final var subscribedListener = SubscribableListener.newForked(l -> initialListener.addListener(l, executor, null));
+        final var andThenListener = initialListener.andThen(executor, null, (l, x) -> fail("should not be called"));
+
+        assertFalse(subscribedListener.isDone());
+        assertFalse(andThenListener.isDone());
+
```
Comment (DiannaHohensee) [inline server/src/test/java/org/elasticsearch/action/support/SubscribableListenerTests.java:625]:
```suggestion

Randomly complete or fail the original subscribable listener. The listener subscribed to the original is going to encounter an Executor exception regardless and overwrite the original success or failure response.
```

**Review by DiannaHohensee — APPROVED** [review]
Some suggestions for clarify, but LGTM 👍

File: server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java:196]: I think I'd interpret that as meaning a subclass of `java.util.concurrent.ExecutionException` which isn't the case for `RejectedExecutionException` or `EsRejectedExecutionException` - the docs for `ExecutionException` say it is "thrown when attempting to retrieve the result of a task that aborted by throwing an exception" which doesn't apply here.

(acknowledging that `RejectedExecutionException` does have the suffix `ExecutionException` but I'd rather call it a rejection exception in this context)

File: server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java (hunk 4 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java:191]: The `Ensure that this {@link SubscribableListener} is always completed using that executor` only applies for the overloads that don't take their own `Executor` argument.

File: server/src/test/java/org/elasticsearch/action/support/SubscribableListenerTests.java (hunk 5 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/action/support/SubscribableListenerTests.java]: Not an `Error` (and `ExecutorExecution` is kinda awkward), proposing `testRejectedExecutionThreading` instead, see d78df289017.

**Pull request merged**